types/wlr_input_method_v2.c

```diff
@@ -108,10 +116,134 @@ static void im_delete_surrounding_text(struct wl_client *client,
+		popup_surface->mapped = true;
+		wlr_signal_emit_safe(&popup_surface->events.map, popup_surface);
+	} else if (!mapped && popup_surface->mapped) {
+		wlr_signal_emit_safe(&popup_surface->events.unmap, popup_surface);
+		popup_surface->mapped = false;
+	}
+}
+
+static void popup_surface_surface_role_commit(struct wlr_surface *surface) {
+	struct wlr_input_popup_surface_v2 *popup_surface = surface->role_data;
+	if (popup_surface == NULL) {
+		return;
+	}
+	popup_surface_set_mapped(popup_surface, wlr_surface_has_buffer(surface)
+		&& popup_surface->input_method->client_active);
+}
+
+static const struct wlr_surface_role input_popup_surface_v2_role = {
+	.name = "zwp_input_popup_surface_v2",
+	.commit = popup_surface_surface_role_commit,
+};
+
+bool wlr_surface_is_input_popup_surface_v2(struct wlr_surface *surface) {
+	return surface->role == &input_popup_surface_v2_role;
+}
+
+static void popup_surface_destroy(
+		struct wlr_input_popup_surface_v2 *popup_surface) {
+	popup_surface_set_mapped(popup_surface, false);
+	wlr_signal_emit_safe(&popup_surface->events.destroy, NULL);
+	wl_list_remove(&popup_surface->surface_destroy.link);
+	wl_list_remove(&popup_surface->link);
+	wl_resource_set_user_data(popup_surface->resource, NULL);
+	free(popup_surface);
+}
+
+static void popup_surface_handle_surface_destroy(struct wl_listener *listener,
+		void *data) {
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		wl_container_of(listener, popup_surface, surface_destroy);
+	popup_surface_destroy(popup_surface);
+}
+
+static void popup_resource_destroy(struct wl_resource *resource) {
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		wl_resource_get_user_data(resource);
+	if (popup_surface == NULL) {
+		return;
+	}
+	popup_surface_destroy(popup_surface);
+}
+
+static void popup_destroy(struct wl_client *client,
+		struct wl_resource *resource) {
+	wl_resource_destroy(resource);
+}
+
+static const struct zwp_input_popup_surface_v2_interface input_popup_impl = {
+	.destroy = popup_destroy,
+};
+
 static void im_get_input_popup_surface(struct wl_client *client,
 		struct wl_resource *resource, uint32_t id,
 		struct wl_resource *surface) {
-	wlr_log(WLR_INFO, "Stub: zwp_input_method_v2::get_input_popup_surface");
+	struct wlr_input_method_v2 *input_method =
+		input_method_from_resource(resource);
+	if (!input_method) {
+		return;
+	}
+
+	struct wl_resource *popup_resource = wl_resource_create(
+		client, &zwp_input_popup_surface_v2_interface,
+		wl_resource_get_version(resource), id);
+	if (!popup_resource) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+
+	struct wlr_input_popup_surface_v2 *popup_surface =
+		calloc(1, sizeof(struct wlr_input_popup_surface_v2));
+	if (!popup_surface) {
+		wl_client_post_no_memory(client);
+		return;
+	}
+	wl_resource_set_implementation(popup_resource, &input_popup_impl,
+		popup_surface, popup_resource_destroy);
+
+	struct wlr_surface *surface_resource = wlr_surface_from_resource(surface);
```

Comment:
Nit: the names are mixed up here, the `wlr_surface` is named `surface_resource` and the `wl_resource` is named `surface`. We usually do it the other way around :P